ethcore/engines/validator-set/src/safe_contract.rs

```diff
@@ -300,6 +383,88 @@ impl ValidatorSet for ValidatorSafeContract {
 			.map(|out| (out, Vec::new()))) // generate no proofs in general
 	}
 
+	fn generate_engine_transactions(&self, _first: bool, header: &Header, caller: &mut SystemCall)
+		-> Result<Vec<(Address, Bytes)>, EthcoreError>
+	{
+		// Skip the rest of the function unless there has been a transition to POSDAO AuRa.
+		if self.posdao_transition.map_or(true, |block_num| header.number() < block_num) {
+			trace!(target: "engine", "Skipping a call to emitInitiateChange");
+			return Ok(Vec::new());
+		}
+		let mut transactions = Vec::new();
+
+		// Create the `InitiateChange` event if necessary.
+		let (data, decoder) = validator_set::functions::emit_initiate_change_callable::call();
+		let emit_initiate_change_callable = caller(self.contract_address, data)
+			.and_then(|x| decoder.decode(&x)
+			.map_err(|x| format!("chain spec bug: could not decode: {:?}", x)))
+			.map_err(EngineError::FailedSystemCall)?;
+		if !emit_initiate_change_callable {
+			trace!(target: "engine", "New block #{} issued ― no need to call emitInitiateChange()", header.number());
+		} else {
+			trace!(target: "engine", "New block issued #{} ― calling emitInitiateChange()", header.number());
+			let (data, _decoder) = validator_set::functions::emit_initiate_change::call();
+			transactions.push((self.contract_address, data));
+		}
+
+		// Retry all pending reports.
+		self.filter_report_queue(header.author(), header.number())?;
```

Comment:
How come `filter_report_queue()` isn't implemented as a member of the `queued_reports` itself? So the line below could be something like `let queued_reports = self.queued_reports.lock().unwrap().filter_reports()` (or something to the same effect)?

Reply:
I extracted the report queue into a new struct; please let me know whether that's what you had in mind. (I put it in a separate commit, so we can revert it if you don't agree.)